elastic/kibana
# Add support for a declarative (via configuration) way to specify Kibana feature overrides

## Summary

This PR extends the features plugin to accept feature definition overrides via Kibana configuration. The functionality is limited to the Serverless offering only. Additionally, the PR updates Kibana serverless configurations to include overrides based on the "simplified feature toggles" proposals discussed with the solution teams.

The configuration might look like this:

```yaml
## Fine-tune the feature privileges.
xpack.features.overrides:
  dashboard:
    privileges:
      ### Dashboard's `All` feature privilege should implicitly
      ### grant `All` access to Maps and Visualize features.
      all.composedOf:
        - feature: "maps"
          privileges: [ "all" ]
        - feature: "visualize"
          privileges: [ "all" ]
    ### All Dashboard sub-feature privileges should be hidden: 
    ### reporting capabilities will be granted via dedicated
    ### Reporting feature and short URL sub-feature privilege
    ### should be granted for both `All` and `Read`.
    subFeatures.privileges:
      download_csv_report.disabled: true
      url_create:
        disabled: true
        includeIn: "read"
  ### Maps feature is disabled since it's automatically granted by Dashboard feature.
  maps.disabled: true
```


## How to test

Log in as the `admin` using SAML and navigate to the `Custom roles` management section to edit role and see tuned role management UX:

<p align="center">
  <img src="https://github.com/elastic/kibana/assets/1713708/ad6e4b07-53bd-4f5a-ae91-66d6534c711a" />
  <img src="https://github.com/elastic/kibana/assets/1713708/8ab4d5a3-f719-42d5-a278-3aee87603c33" />
</p>

![image](https://github.com/elastic/kibana/assets/1713708/5e27a49b-4382-4a91-bb85-eca929a27961)

### Search project
```bash
yarn es serverless --projectType=es --ssl -E xpack.security.authc.native_roles.enabled=true
yarn start --serverless=es --ssl --xpack.security.roleManagementEnabled=true
```

Refer to the proposal document, `config/serverless.yml`, and `config/serverless.es.yml` in this PR to see the specific changes made for your project type:

![image](https://github.com/elastic/kibana/assets/1713708/9f9d0341-32a1-4258-be3b-d3a809f5bacc)

Create a custom `custom-search` role and re-login as the user with this role to test your project type (you need to manually type role name if the role selector):

<p align="center">
  <img src="https://github.com/elastic/kibana/assets/1713708/5088320b-3cc8-4de9-984c-d70fc6277659" />
</p>

### Observability project
```bash
yarn es serverless --projectType=oblt --ssl -E xpack.security.authc.native_roles.enabled=true
yarn start --serverless=oblt --ssl --xpack.security.roleManagementEnabled=true
```

Refer to the proposal document, `config/serverless.yml`, and `config/serverless.oblt.yml` in this PR to see the specific changes made for your project type:

![image](https://github.com/elastic/kibana/assets/1713708/1d2b360a-24ab-47f7-ac9b-8ad944949c32)

Create a custom `custom-o11y` role and re-login as the user with this role to test your project type (you need to manually type role name if the role selector):

<p align="center">
  <img src="https://github.com/elastic/kibana/assets/1713708/110572b1-f08a-4427-a687-5c2e0240a36b" />
</p>

### Security project
```bash
yarn es serverless --projectType=security --ssl -E xpack.security.authc.native_roles.enabled=true
yarn start --serverless=security --ssl --xpack.security.roleManagementEnabled=true
```

Refer to the proposal document, `config/serverless.yml`, and `config/serverless.security.yml` in this PR to see the specific changes made for your project type:

![image](https://github.com/elastic/kibana/assets/1713708/2dbca002-59f1-44f0-9ab2-1dd205e48da8)

Create a custom `custom-security` role and re-login as the user with this role to test your project type (you need to manually type role name if the role selector):

<p align="center">
  <img src="https://github.com/elastic/kibana/assets/1713708/2bec6ae2-8d19-4142-a479-9a81bc1fca14" />
</p>

__Fixes: https://github.com/elastic/kibana/issues/178963__
